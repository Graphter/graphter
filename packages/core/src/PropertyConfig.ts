import { PropertyValidator } from "./PropertyValidator";

export interface PropertyConfig {
  id: string
  name?: string
  description?: string
  type: string
  listItemType?: string
  default?: any | Function
  validation?: PropertyValidator | Array<PropertyValidator>
}