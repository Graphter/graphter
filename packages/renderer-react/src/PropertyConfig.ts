import {ValidationConfig} from "./ValidationConfig";

export interface PropertyConfig {
  id: string,
  name?: string,
  description?: string,
  type: string,
  validations?: Array<ValidationConfig>,
  default?: any | Function
}