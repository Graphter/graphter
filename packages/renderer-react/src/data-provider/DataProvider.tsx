import React, {createContext, useContext} from 'react';
import { Service } from "microo-core";

interface DataProviderProps {
  service: Service,
  children: any
}

const Context = createContext<Service | null>(null);

export function useService(): Service {
  const service = useContext(Context);
  if(!service) throw new Error(`Couldn't find API service. Make sure you've declared a DataProvider and passed it a valid service.`);
  return service;
}

export default function DataProvider({ service, children }: DataProviderProps){
  return (
    <Context.Provider value={service}>
      {children}
    </Context.Provider>
  );
}