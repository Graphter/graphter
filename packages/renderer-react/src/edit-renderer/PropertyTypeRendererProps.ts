import { DisplayTypeRendererProps } from "./DisplayTypeRendererProps";
import { EditDisplayConfig, ErrorRendererProps, ModelConfig, PropertyConfig, ValidationResult } from "microo-core";
import { RecoilState } from "recoil";
import { ComponentType } from "react";

export interface PropertyTypeRendererProps extends DisplayTypeRendererProps {
  propertyConfig: PropertyConfig
  displayConfig: EditDisplayConfig
  modelConfig: ModelConfig
  propData: any
  modelData: any
  setPropDataValue: (value: any) => void
  setModelDataValue: (value: any) => void
  validationResults?: Array<ValidationResult>
  errorRenderer: ComponentType<ErrorRendererProps> | null
  displayTypeRenderers: { [typeId: string]: ComponentType<DisplayTypeRendererProps> }
  propertyTypeRenderers: { [typeId: string]: ComponentType<PropertyTypeRendererProps> }
}