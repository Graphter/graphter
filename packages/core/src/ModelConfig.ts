import {PropertyConfig} from "./PropertyConfig";
import { ListDisplayConfig } from "./ListDisplayConfig";
import { EditDisplayConfig } from "./EditDisplayConfig";
import { ModelValidator } from "./ModelValidator";

export interface ModelConfig {
  id: string
  name: string
  description?: string
  properties: Array<PropertyConfig>
  identityPath?: Array<string>
  display?: {
    list?: ListDisplayConfig
    edit?: Array<EditDisplayConfig | string>
  },
  validation?: ModelValidator | Array<ModelValidator>
}