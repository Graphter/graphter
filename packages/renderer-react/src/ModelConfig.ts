import { ReactNode } from "react";
import {PropertyConfig} from "./PropertyConfig";
import ListDisplayConfig from "./ListDisplayConfig";
import EditDisplayConfig from "./EditDisplayConfig";

export default interface ModelConfig {
  id: string,
  name: string,
  description?: string,
  properties: Array<PropertyConfig>,
  identityPath?: Array<string>,
  display?: {
    list?: ListDisplayConfig,
    edit?: Array<EditDisplayConfig | string>
  }
}