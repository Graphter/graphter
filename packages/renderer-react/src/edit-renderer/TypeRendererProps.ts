import {PropertyConfig} from "../PropertyConfig";
import EditDisplayConfig from "../EditDisplayConfig";
import {ComponentType} from "react";

export interface TypeRendererProps {
  data: any,
  displayConfig: EditDisplayConfig,
  renderChildren: (displayConfig: Array<EditDisplayConfig | string>) => ComponentType<any>,
  onChange?: (data: any) => void
}