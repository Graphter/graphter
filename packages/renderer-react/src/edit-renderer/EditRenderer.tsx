import React, {Component, ComponentType, useEffect, useState} from "react";
import s from './EditRenderer.pcss';
import ModelConfig from "../ModelConfig";
import {expand, isPropertyConfig} from "../util/editDisplayConfig";
import {isEmpty} from "../util/id";
import EditDisplayConfig from "../EditDisplayConfig";
import {getProp, queryProp} from "../util/propertyConfig";
import {createNewInstance} from "../util/model";
import {TypeRendererProps} from "./TypeRendererProps";
import {PropertyTypeRendererProps} from "./PropertyTypeRendererProps";
import ErrorBoundary from "../error-boundary";
import {useService} from "../data-provider/DataProvider";

export interface EditRendererProps {
  config: ModelConfig,
  id?: string | number,
  typeRenderers?: { [typeId: string]: ComponentType<TypeRendererProps> }
  propertyTypeRenderers?: { [typeId: string]: ComponentType<PropertyTypeRendererProps> },
  errorRenderer?: ComponentType<{ err: Error }>
}

export default function EditRenderer({ config, id, typeRenderers, propertyTypeRenderers, errorRenderer }: EditRendererProps){
  const ErrorDisplayComponent: ComponentType<{ err: Error }> = errorRenderer || (({ err }) => <div data-testid='error'>error: {err.message}</div>);

  if(!config) return <ErrorDisplayComponent err={new Error('Model configuration is required')} />;

  const service = useService();

  const [ editingModel, setEditingModel ] = useState<any>()
  const [ loading, setLoading ] = useState(true);
  const [ error, setError ] = useState<Error>();

  useEffect(() => {
    (async () => {
      if(isEmpty(id)) {
        setEditingModel(createNewInstance(config));
        setLoading(false);
      } else {
        try {
          const getResult = await service.get(config.id, id);
          setLoading(false);
          if (!getResult.item) {
            setError(new Error(`Couldn't find a ${config.name} with ID '${id}'`));
            return;
          }
          setEditingModel(getResult.item);
        } catch(err){
          setLoading(false);
          setError(new Error(`There was a problem loading that ${config.name}: ${err.message}`));
        }
      }
    })()
  }, [ id ]);

  return (
    <div className={s.editRenderer}>

      {error && <ErrorDisplayComponent err={error} />}

      {loading && <div className={s.editRenderer} data-testid='loading'>loading...</div>}

      <form onSubmit={() => service.save(config.id, editingModel)} data-testid='form'>
        <ErrorBoundary errorRenderer={ErrorDisplayComponent}>
          {editingModel && renderFromDisplayConfig(config.display?.edit)}
        </ErrorBoundary>
        <button type='submit' data-testid='save'>Save</button>

      </form>

    </div>
  );

  function renderFromDisplayConfig(displayConfig?: Array<EditDisplayConfig | string>): any{
    try {
      const expandedDisplayConfig = expand(displayConfig, config.properties);

      return expandedDisplayConfig.map((itemDisplayConfig, i) => {
        return <ErrorBoundary key={i} errorRenderer={ErrorDisplayComponent}>
          {(() => {
            if (!itemDisplayConfig.typeRenderer) {
              throw new Error(`Somehow ${itemDisplayConfig.type} ended up without a type Renderer. Sounds like a problem with microo ðŸ¥º`);
            }

            if (isPropertyConfig(itemDisplayConfig)) {
              const matchingProp = queryProp(itemDisplayConfig.options.property, config.properties);
              if (!matchingProp) {
                throw new Error(`Couldn't find prop matching '${itemDisplayConfig.type}' display config`);
              }
              const TypeRenderer = propertyTypeRenderers && propertyTypeRenderers[itemDisplayConfig.typeRenderer || matchingProp.type];
              if (!TypeRenderer) {
                throw new Error(`No property type renderer for type '${itemDisplayConfig.typeRenderer || matchingProp.type}' of '${matchingProp.name}'. Registered property type renderers: [${Object.keys(propertyTypeRenderers || {}).join(`, `)}]`);
              }
              return <TypeRenderer
                data={editingModel}
                propertyConfig={getProp(itemDisplayConfig.options.property, config.properties)}
                displayConfig={itemDisplayConfig}
                onChange={(data: any) => {
                  setEditingModel(data);
                }}
                renderChildren={renderFromDisplayConfig}/>
            } else {
              const TypeRenderer = typeRenderers && typeRenderers[itemDisplayConfig.type];
              if (!TypeRenderer) {
                throw new Error(`No type renderer for display type '${itemDisplayConfig.type}'. Registered type renderers: [${Object.keys(typeRenderers || {}).join(`, `)}]`);
              }
              return <TypeRenderer
                data={editingModel}
                displayConfig={itemDisplayConfig}
                renderChildren={renderFromDisplayConfig}/>
            }
          })()}
        </ErrorBoundary>
      });
    } catch(err){
      return <ErrorDisplayComponent err={err} />
    }
  }

}

