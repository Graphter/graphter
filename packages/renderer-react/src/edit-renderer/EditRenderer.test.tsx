import React, {ComponentType} from "react";
import EditRenderer from "./EditRenderer";
import { act, render, within, fireEvent } from '@testing-library/react';
import { screen } from '@testing-library/dom'
import '@testing-library/jest-dom/extend-expect'
import {PropertyConfig} from "../PropertyConfig";
import ModelConfig from "../ModelConfig";
import {prop} from "ramda";
import {TypeRendererProps} from "./TypeRendererProps";
import {PropertyTypeRendererProps} from "./PropertyTypeRendererProps";
import DataProvider from "../data-provider";

describe(`<EditRenderer />`, () => {
  let baseConfig: ModelConfig,
    propertyTypeRenderers: { [typeId: string]: ComponentType<PropertyTypeRendererProps> },
    mockService: any;

  beforeEach(() => {
    baseConfig = {
      id: 'page',
      name: 'Page',
      properties: [
        {
          id: 'id',
          name: 'ID',
          type: 'string',
          default: () => {
            return 'a-generated-id'
          }
        },
        {
          id: 'title',
          name: 'Title',
          type: 'string',
          default: 'Untitled'
        }
      ]
    };
    propertyTypeRenderers = {
      'string': ( { data, propertyConfig, onChange }: PropertyTypeRendererProps ) => (
        <div data-testid={`property-${propertyConfig.id}`}>{data[propertyConfig.id]}</div>
      )
    };

    mockService = {
      list: jest.fn().mockResolvedValue({
        items: [],
        count: 10,
        skip: 0,
        take: 10
      }),
      get: jest.fn().mockResolvedValue({
        item: {},
        version: 1,
        created: new Date(),
        updated: new Date(),
        deleted: false
      }),
      save: jest.fn().mockResolvedValue({
        item: {},
        version: 2,
        updated: new Date()
      }),
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe(`when editing a new item`, () => {
    it(`should initialise the model instance with default values`, async () => {
      baseConfig.properties[0].default = () => {
        return 'a-generated-id'
      };
      baseConfig.properties[1].default = 'Untitled';

      await act(async() => {
        render(
          <DataProvider service={mockService}>
            <EditRenderer config={baseConfig} propertyTypeRenderers={propertyTypeRenderers}/>
          </DataProvider>);
      });

      const idProp = screen.getByTestId('property-id');
      expect(idProp).toHaveTextContent('a-generated-id');
      const titleProp = screen.getByTestId('property-title');
      expect(titleProp).toHaveTextContent('Untitled');
    });
  });
  describe(`when editing an existing item`, () => {
    it(`should load the item to edit`, async () => {
      await act(async() => {
        render(
          <DataProvider service={mockService}>
            <EditRenderer config={baseConfig} id={123} propertyTypeRenderers={propertyTypeRenderers} />
          </DataProvider>
          );
      });
      expect(mockService.get).toHaveBeenCalledWith('page', 123);
    });
    it(`should display an appropriate error when no model configuration is provided`, async () => {
      await act(async() => {
        render(
          <DataProvider service={mockService}>
            // @ts-ignore
            <EditRenderer config={null} id={'some-id'} propertyTypeRenderers={propertyTypeRenderers} />
          </DataProvider>);
      });
      const error = screen.getByTestId('error');
      expect(error.textContent).toMatchSnapshot();
    });
    it(`should display an appropriate error when it can't find item with the specified ID`, async () => {
      mockService.get.mockResolvedValue({
        item: null
      });
      await act(async() => {
        render(
          <DataProvider service={mockService}>
            <EditRenderer config={baseConfig} id={'some-id'} propertyTypeRenderers={propertyTypeRenderers} />
          </DataProvider>);
      });
      const error = screen.getByTestId('error');
      expect(error.textContent).toMatchSnapshot();
    });
    it(`should display an appropriate error when service can't load data`, async () => {
      mockService.get.mockRejectedValue({ message: 'some error message' });
      await act(async() => {
        render(
          <DataProvider service={mockService}>
            <EditRenderer config={baseConfig} id={'some-id'} propertyTypeRenderers={propertyTypeRenderers} />
          </DataProvider>);
      });
      const error = screen.getByTestId('error');
      expect(error.textContent).toMatchSnapshot();
    });
  });
  it(`should render the UI to edit each property`, async () => {
    mockService.get.mockResolvedValue({
      item: {
        id: 'some-id',
        title: 'Some Page Title'
      },
      version: 1,
      created: new Date(),
      updated: new Date(),
      deleted: false
    })
    await act(async() => {
      render(
        <DataProvider service={mockService}>
          <EditRenderer config={baseConfig} id={'some-id'} propertyTypeRenderers={propertyTypeRenderers} />
        </DataProvider>);
    });
    const idProp = screen.getByTestId('property-id');
    expect(idProp).toHaveTextContent('some-id');
    const titleProp = screen.getByTestId('property-title');
    expect(titleProp).toHaveTextContent('Some Page Title');
  });
  it(`should show loading UI to begin with`, async () => {
    await act(async() => {
      render(
        <DataProvider service={mockService}>
          <EditRenderer config={baseConfig} id={'some-id'} propertyTypeRenderers={propertyTypeRenderers} />
        </DataProvider>);
      const loading = screen.queryByTestId('loading');
      expect(loading?.textContent).toMatchSnapshot();
    });
  });
  it(`should hide the loading UI once item has been loaded`, async () => {
    await act(async() => {
      render(
        <DataProvider service={mockService}>
          <EditRenderer config={baseConfig} id={'some-id'} propertyTypeRenderers={propertyTypeRenderers} />
        </DataProvider>);
    });
    const loading = screen.queryByTestId('loading');
    expect(loading).toBeNull();
  });
  it(`should render the UI for non-property display config`, async () => {
    baseConfig.display = {
      edit: [
        {
          type: 'column',
          children: [
            { type: 'property', options: { property: 'title' }}
          ]
        }
      ]
    };
    await act(async () => {
      render(
        <DataProvider service={mockService}>
          <EditRenderer
            config={baseConfig}
            propertyTypeRenderers={propertyTypeRenderers}
            typeRenderers={{
              'column': ({ data, displayConfig, renderChildren}) => (
                <div data-testid='col'>{displayConfig.children && renderChildren(displayConfig.children)}</div>
              )
            }}/>
        </DataProvider>);
    });
    const column = screen.getByTestId('col');
    expect(column).toMatchSnapshot();
  });
  it(`should display an appropriate error when a property display config doesn't match a property`, async () => {
    mockService.get.mockResolvedValue({
      item: {
        id: 'some-id',
        title: 'Some Page Title'
      },
      version: 1,
      created: new Date(),
      updated: new Date(),
      deleted: false
    })
    baseConfig.display = {
      edit: [
        'some-weird-unknown-prop'
      ]
    }
    await act(async() => {
      render(
        <DataProvider service={mockService}>
          <EditRenderer config={baseConfig} id={'some-id'} propertyTypeRenderers={propertyTypeRenderers} />
        </DataProvider>);
    });
    const error = screen.getByTestId('error');
    expect(error?.textContent).toMatchSnapshot();
  });
  it(`should display an appropriate error when it can't find the correct property type renderer`, async () => {
    mockService.get.mockResolvedValue({
      item: {
        id: 'some-id',
        title: 'Some Page Title'
      },
      version: 1,
      created: new Date(),
      updated: new Date(),
      deleted: false
    })
    baseConfig.properties.push({
      id: 'prop-of-unknown-type',
      name: 'Prop of unknown type',
      type: 'rainbow'
    })
    await act(async() => {
      render(
        <DataProvider service={mockService}>
          <EditRenderer config={baseConfig} id={'some-id'} propertyTypeRenderers={propertyTypeRenderers} />
        </DataProvider>);
    });
    const error = screen.getByTestId('error');
    expect(error?.textContent).toMatchSnapshot();
  });
  it(`should display an appropriate error when it can't find the correct non-property type renderer`, async () => {
    mockService.get.mockResolvedValue({
      item: {
        id: 'some-id',
        title: 'Some Page Title'
      },
      version: 1,
      created: new Date(),
      updated: new Date(),
      deleted: false
    });
    baseConfig.display = {
      edit: [
        {
          type: 'column'
        }
      ]
    }
    await act(async() => {
      render(
        <DataProvider service={mockService}>
          <EditRenderer config={baseConfig} id={'some-id'} propertyTypeRenderers={propertyTypeRenderers} />
        </DataProvider>);
    });
    const error = screen.getByTestId('error');
    expect(error?.textContent).toMatchSnapshot();
  });
  it(`should save the model instance when the form submits (either with enter key or clicking Save)`, async () => {
    mockService.get.mockResolvedValue({
      item: {
        id: 'some-id',
        title: 'Some Page Title'
      },
      version: 1,
      created: new Date(),
      updated: new Date(),
      deleted: false
    });
    await act(async() => {
      render(
        <DataProvider service={mockService}>
          <EditRenderer config={baseConfig} id={'some-id'} propertyTypeRenderers={propertyTypeRenderers} />
        </DataProvider>);
    });
    fireEvent.submit(screen.getByTestId('form'));
    expect(mockService.save).toHaveBeenCalledWith(
      'page',
      {
        id: 'some-id',
        title: 'Some Page Title'
      });
  });
});

export {};