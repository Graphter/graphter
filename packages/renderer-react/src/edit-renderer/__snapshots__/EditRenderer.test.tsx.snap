// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditRenderer /> should display an appropriate error when a property display config doesn't match a property 1`] = `"error: Couldn't find prop matching 'some-weird-unknown-prop' from layout config. Check the property exists."`;

exports[`<EditRenderer /> should display an appropriate error when it can't find the correct non-property type renderer 1`] = `"error: No type renderer for display type 'column'. Registered type renderers: []"`;

exports[`<EditRenderer /> should display an appropriate error when it can't find the correct property type renderer 1`] = `"error: No property type renderer for type 'rainbow' of 'Prop of unknown type'. Registered property type renderers: [string]"`;

exports[`<EditRenderer /> should hide the loading UI once item has been loaded 1`] = `"loading..."`;

exports[`<EditRenderer /> should render the UI for non-property display config 1`] = `
<div
  data-testid="col"
>
  <div
    data-testid="property-title"
  >
    Untitled
  </div>
</div>
`;

exports[`<EditRenderer /> should show loading UI to begin with 1`] = `"loading..."`;

exports[`<EditRenderer /> when editing an existing item should display an appropriate error when it can't find item with the specified ID 1`] = `"error: Couldn't find a Page with ID 'some-id'"`;

exports[`<EditRenderer /> when editing an existing item should display an appropriate error when no model configuration is provided 1`] = `"error: Model configuration is required"`;

exports[`<EditRenderer /> when editing an existing item should display an appropriate error when service can't load data 1`] = `"error: There was a problem loading that Page: some error message"`;
